_posts/2020-12-08-avro-native.adoc

```diff
@@ -0,0 +1,102 @@
+author: cescoffier
+---
+
+A few months back, I blogged about https://quarkus.io/blog/kafka-avro/[Quarkus, Kafka, and Avro] (using the Apicurio schema registry).
+
+We developed a simple application, receives HTTP requests, writes Kafka records, and reads them from Kafka.
+It uses Avro to serialize and deserialize the records.
+
+image:/assets/images/posts/kafka-avro/architecture.png[]
+
+At that time, the https://github.com/cescoffier/quarkus-kafka-and-avro[application] was working in native mode.
+
+With Quarkus 1.10.2, it's now working in native mode!
+
+== Why didn't it work?
+
+The native compiler does not support everything that is possible to do in Java.
+There are cases not yet supported, and Avro is using some of these unsupported constructs.
+Fortunately, we can implement _substitutions_ (to replace unsupported constructs in the code directly) and use Quarkus extensions to configure the compiler correctly.
+As a result, your application does not have to configure anything, "it just works™️".
+
+Let's come back to Avro.
+Avro uses https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.html[method handles] that are not supported by the GraalVM native compiler.
+
+To workaround the unsupported constructs used by Avro, we implemented a set of https://github.com/quarkusio/quarkus/blob/master/extensions/avro/runtime/src/main/java/io/quarkus/avro/graal/AvroSubstitutions.java[substitutions].
+We replaced method handles with reflection.
+
+The `GenericDatumReader` also needs a bit of work as it touches threads at build time.
+
+Finally, in the Quarkus Avro Processor (part of the Quarkus extension), we register for reflection all the classes annotated with `@AvroGenerated`, as we are using reflection to create new instances of them.
+
+== Show me!
+
+Check out the https://github.com/cescoffier/quarkus-kafka-and-avro[code] of the application, and make sure you have GraalVM installed.
+Then, build the native executable using:
+
+[source, bash]
+----
+mvn package -Pnative
+----
+
+Grab a coffee or tea, as it may take a few minutes
+
+Start the Kafka broker and the schema registry using:
+
+[source, bash]
+----
+docker-compose up -d
+----
+
+Finally, run the application with `./target/kafka-and-avro-1.0.0-SNAPSHOT-runner`.
+
+Send some movies to the application:
+
+[source, bash]
+----
+curl --header "Content-Type: application/json" \
+--request POST \
+--data '{"title":"The Shawshank Redemption","year":1994}' \
+http://localhost:8080/movies
+
+curl --header "Content-Type: application/json" \
+--request POST \
+--data '{"title":"The Godfather","year":1972}' \
+http://localhost:8080/movies
+
+curl --header "Content-Type: application/json" \
+--request POST \
+--data '{"title":"The Dark Knight","year":2008}' \
+http://localhost:8080/movies
+
+curl --header "Content-Type: application/json" \
+--request POST \
+--data '{"title":"12 Angry Men","year":1957}' \
+http://localhost:8080/movies
+----
+
+Check they are processed successfully:
```

Comment:
Does this check require a command, or is it just the docker compose terminal output?

Might be good to clarify where you see these entries